packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+// @ts-check
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        const handler = wasFulfilled ? watcher.onFulfilled : watcher.onRejected;
+        if (handler) {
+          handler(value, ...args);
+        }
```

Comment:
handler is user provided and can throw, which shouldn't impact other watchers.

```suggestion
        try {
          handler(value, ...args);
        } catch(e) {}
```

But I would actually prefer if each handler was called on a clean stack:

```suggestion
        Promise.resolve().then(() => { handler(value, ...args); });
```